### ISSUE TITLE
ARM M1 compiling

### ISSUE THREAD

**Issue by juanfal** [quicksilver/Quicksilver#2543]
M1 is there.  It's the greatest increase in speed in one step I've ever seen.  Quicksilver is, by far, the most useful utility made for OSX ever.  QS has several issues on Big Sur, specifically the copy-paste history… but apart from fixing bugs, it'd be great to know if is there any plans to recompile it for ARM

**Comment by skurfer** [comment]
I can get it to compile in Xcode 12, so it should be possible in theory, but there are a lot of problems with the resulting application that you don’t see in the current version (built with Xcode 9), so it isn’t something we would want to release as is.

I would like to fix the bugs, but I don’t know if I’ll have time, or if I would even know how. I can only speak for myself though. It’s an open source project that anyone could contribute to.

**Comment by pasiaj** [comment]
@skurfer, I'd love to start helping out on M1 with a long term goal of having a branch that can be compiled on latest Xcode without modifications and then start encouraging others to jump on. 

If someone was to help me get started with getting the source to compile on Xcode 12.4. 

Right now the build fails on all build targets with:
"unable to find sdk 'macosx10.13'"

**Comment by juanfal** [comment]
I did try it too, and started having the same issue.  Not being involved in any other similar project now, sorry I can't help here.

**Comment by skurfer** [comment]
@pasiaj I’ve just committed some changes that allow me to build it under Xcode 12. You’ll want to pick Quicksilver Distribution as the target if it’s not already selected. Let me know if you run into problems. I’ve sent you an invite for Slack as well.

**Comment by totallyGreg** [comment]
Pulled latest check-in
❯ git hist
* 4374aea4 update the Xcode project version (HEAD -> master, origin/master, origin/HEAD) (19 hours ago) [Rob McBroom]
Updated the submodules
❯ git submodule update
Submodule path 'Quicksilver/Code-External/DisableSubviews': checked out 'f569e60e9e46220bce701dcfe7b92f4d8abe01ad'
Submodule path 'Quicksilver/Code-External/FMDB': checked out '6c0ff9c423e8b476c63f11413790443c3fe62f16'
Submodule path 'Quicksilver/Code-External/LaunchAtLoginController': checked out 'da5db7a0a20e76f83a10f7ecd34c81d1def7a64d'
Submodule path 'Quicksilver/Code-External/VDKQueue': checked out '503c187d0a418d0af8c90a6216435b3052b0b223'

and Xcode 12.4 Build fails with the same error 
Showing All Messages
unable to find sdk 'macosx10.13'

Currently I'm building on on my Intel Laptop, but I have an M1 mini I plan to work on as well.  Very much want to contribute back to the Quicksilver love I have enjoyed over the last 2 decades.

**Comment by skurfer** [comment]
> Xcode 12.4 Build fails with the same error
> Showing All Messages
> unable to find sdk ‘macosx10.13'

ee1ef2c7a5eb12c250d5a995eb9fcd953904add3 definitely changes the SDK. Try cleaning with ⇧⌘K or even quitting Xcode, deleting `/tmp/QS`, and starting it again. You might also want to go through all of the targets and make sure none of them have manually specified a Base SDK. (They should all be getting it from the `xcconfig` file.)

**Comment by n8henrie** [comment]
Thanks for the recent Xcode 12 compatibility commit! I tried getting it to build on my M1 mac a few weeks ago and just am not familiar enough with Xcode.

At 4374aea4 and the tiny patch below to remove `VALID_ARCHS`, it builds first try on my M1 Mac and runs on silicon!

<img width="1150" alt="Screen Shot 2021-04-03 at 14 54 46" src="https://user-images.githubusercontent.com/1234956/113491419-dfbdf580-948d-11eb-808e-aef32f29a41f.png">




```
diff --git a/Quicksilver/Configuration/Common.xcconfig b/Quicksilver/Configuration/Common.xcconfig
index d65e9a1f..0bca8768 100644
--- a/Quicksilver/Configuration/Common.xcconfig
+++ b/Quicksilver/Configuration/Common.xcconfig
@@ -1,7 +1,5 @@
 #include "Developer.xcconfig"
 
-VALID_ARCHS = x86_64
-
 ARCHS = $(ARCHS_STANDARD_64_BIT)
 
 SDKROOT = macosx
```

**Comment by skurfer** [comment]
Thanks @n8henrie. I’ve pushed that change as well.

**Comment by n8henrie** [comment]
Trying to figure out the process for rebuilding plugins for Apple Silicon. Not having much luck -- can't find the header files.

Re: [setting `QSFrameworks` in `Source Trees`](https://github.com/quicksilver/Shelf-qsplugin#before-building) -- doesn't look like this option exists in XCode anymore. Tried setting it to `/Applications/Quicksilver.app/Contents/Frameworks` in [Custom Paths](https://stackoverflow.com/a/57712029/1588795) with no luck, though I can see the frameworks there. Tried using `/tmp/QS/build/Release` (where I also find e.g. `QSFoundation.framework`) with no luck.

Any ideas?

Should I start a separate issue for plugins?

**Comment by n8henrie** [comment]
Also just thought I'd share a couple scripts I whipped up to speed up recompiling plugins.

In a new / temporary directory, create a file named `urls.txt` with the github URL to plugins of interest, one per line:

```plaintext
https://github.com/quicksilver/Bezel-qsplugin
https://github.com/quicksilver/BezelClassic-qsplugin
https://github.com/quicksilver/CommandLineTool-qsplugin
https://github.com/quicksilver/Compression-qsplugin
https://github.com/quicksilver/DiskImage-qsplugin
https://github.com/quicksilver/ExtraScripts-qsplugin
https://github.com/quicksilver/Firefox-qsplugin
https://github.com/quicksilver/ImageManipulation-qsplugin
https://github.com/quicksilver/MiniBezel-qsplugin
https://github.com/quicksilver/NotificationHub-qsplugin
https://github.com/quicksilver/com.apple.AddressBook-qsplugin
https://github.com/quicksilver/com.apple.Automator-qsplugin
https://github.com/quicksilver/com.apple.Mail-qsplugin
https://github.com/quicksilver/com.google.mail-qsplugin
https://github.com/quicksilver/elements.hfsattributes-qsplugin.git
https://github.com/quicksilver/elements.support.mail-qsplugin
https://github.com/quicksilver/elements.trigger.event-qsplugin
https://github.com/quicksilver/iCal-qsplugin
https://github.com/quicksilver/iPhoto-qsplugin
https://github.com/quicksilver/iTunes-qsplugin
```

Then run this:

```bash
#!/bin/bash

set -x

build_parallel() {
  find . -mindepth 1 -maxdepth 2 -name '*.xcodeproj' -print0 |
    xargs -0 -I{} xcodebuild -configuration Release -project {}
}

build() {
  find . -mindepth 1 -maxdepth 2 -name '*.xcodeproj' -exec \
    bash -c 'xcodebuild -configuration Release -project "$1" || echo "$1" >> failed.txt' \
    _ {} \;
}

main() {
  # Download in parallel
  while read -r url; do
    git clone "${url}" &
  done < urls.txt
  wait

  build
}
main
```

Feel free to run `build_parallel` instead of `build` the first time, maybe it will speed things up. I ended up with errors in roughly 25% of the plugins, which required manual intervention. The most common error I ran into was a required submodule (fix: `git submodule update --init --recursive`), so you could probably use the `--recurse-submodules` flag to `git clone`. Also had a few files that needed to be converted to UTF8.

I haven't yet figured out how to fix several of them:
- lots of issues with [MailCore](https://github.com/MailCore/MailCore) and the mail plugins)
- FireFox plugin wouldn't even try
- iPhoto plugin is obviously outdated
- Disk Image plugin -- a bunch of missing config files? Maybe just too old to load in modern Xcode

Anyway, I had good luck with more of them than I expected.

**Comment by skurfer** [comment]
Thanks for your work on this @n8henrie.

Yeah, we should start releasing universal versions of the plug-ins first so when the app is updated, they’ll just work. (I knew the Mail plug-in was going to be a problem.) The tricky part is bumping the version and having that match what’s in the repo. It might depend on where `bltrversion` is in a particular plug-in, and at some point Xcode reversed whether it would run before or after building, so I always just end up manually checking and fiddling with it.

**Comment by n8henrie** [comment]
I suppose we could start uploading the universal plugins even before the universal Quicksilver is ready, right?

Is the process for preparing a release build of a plugin any different than building a release build locally? E.g. do I need some kind of signing identity or to strip my personal metadata from anywhere?

**Comment by skurfer** [comment]
> I suppose we could start uploading the universal plugins even before the universal Quicksilver is ready, right?

Yes.

> Is the process for preparing a release build of a plugin any different than building a release build locally? E.g. do I need some kind of signing identity or to strip my personal metadata from anywhere?

We’ve never signed the plug-ins, and I’m not aware of anything else in the built product that identifies the builder.

I don’t have much time to deal with this during the week right now, but I’ll try to set aside some time this weekend to get you more details.

**Comment by n8henrie** [comment]
Sounds good. I've been using the clipboard plugin pretty heavily, and it seems to be working fine. It would probably make sense for me to upload somewhere to have someone test, although it looks like [others should be able to build universal versions as well](https://developer.apple.com/documentation/apple-silicon/building-a-universal-macos-binary):

> You can build a universal binary on either an Apple silicon or Intel-based Mac computer, but you cannot debug the arm64 slice of your binary on an Intel-based Mac computer. It’s possible to debug both slices of a universal binary on Apple silicon, but you must run the x86_64 slice under Rosetta translation.

I had previously thought you needed an ARM mac to build.

FWIW, I just built a universal build of Quicksilver:

```
$ lipo -archs /tmp/QS/build/Release/Quicksilver.app/Contents/MacOS/Quicksilver
x86_64 arm64
```

And ran it on my wife's older Intel MBA; it launched and could browse and select a file, but repeatedly crashed upon *dismissing* the QS window:

```plaintext
Application Specific Information:
*** Terminating app due to uncaught exception 'NSInvalidArgumentException', reason: '-[NSTaggedPointerString frame]: unrecognized selector sent to instance 0x3f847ae147ae147b'
```

**Comment by skurfer** [comment]
The crash probably has more to do with building with the new SDK than anything else. If I just build an Intel only version with Xcode 12, it crashes all the time. Obviously, we *have* to build with Xcode 12 to get M1 support, so we’re going to have to track down and fix these crashes eventually.

**Comment by ghulands** [comment]
@n8henrie and @skurfer: Any progress made since June?

**Comment by n8henrie** [comment]
Thanks for the bump.

Sorry, I'm still really new to this, and I don't really understand the release process.

I've been running a self-compiled universal binary for a couple months, and it works pretty darn well. There have been a few minor issues that oddly seem to be getting better over time for no apparent reason (e.g. issues with the clipboard, issues with proxy objects like `Latest Download`, initially would hang and stall for a few seconds, now are perfectly smooth).

It might be reasonable to cut an alpha release to see what kind of issues others run into, but the problem is that so many plugins also need to be recompiled and should probably also be considered alpha for a while. I don't know the system well enough to know if there's a way for an alpha QS to automatically select on the appropriately compiled plugins.

We could cut an alpha release of QS, but if none (or few) of the plugins worked, it wouldn't do people much good.

@skurfer -- what is the release process with plugins? How do they get uploaded to qsapp.com (I assume that's where the binaries are hosted)? Perhaps we could look at the shelf and clipboard plugins (seem to be working pretty well for me, other options might be calculator, contacts), and if we can test out the universals and they continue working well on x86_64, you could walk me through releasing them? Once some of the core / most popular plugins are ready, that would pave the way for QS itself.

Also, in another thread I mentioned that GitHub Actions doesn't yet support MacOS 11, but once they do, perhaps we should look into a CI automated release system that automatically compiles universal binaries in CI, and change the QS logic to download the plugins from there?

**Comment by skurfer** [comment]
As we’ve talked about, we should start by releasing plug-ins. I’ll reach out to you directly via e-mail about it.

**Comment by jonathanalevi** [comment]
Glad to see someone is working on this, as it would be a massive disappointment if Apple Silicon is what finally kills QuickSilver after years of moderate neglect. 

Hope you all will share you updated builds, since as of today, it's still saying quicksilver "will not be compatible with future versions of macOS"...

**Comment by n8henrie** [comment]
We're working on it.

If anybody wants to give a shot to my first attempt: <https://github.com/n8henrie/Quicksilver/releases/download/1.6.1/Quicksilver.1.6.1.dmg> -- note that this is not yet merged into quicksilver/Quicksilver, coming from my fork while I work out bugs in the GitHub Actions setup.

**Comment by skurfer** [comment]
> it's still saying quicksilver "will not be compatible with future versions of macOS"...

That’s because it runs Python 2 for some things, not because it’s built for Intel.

**Comment by n8henrie** [comment]
> it's still saying quicksilver "will not be compatible with future versions of macOS"

@jonathanalevi Where do you see this? Is it just on installation?

**Comment by skurfer** [comment]
This most likely happens when right-arrowing into a URL, which will run `QSURLExtractor.py` via Python 2. I can’t think of any other part of QS that would run Python.

**Comment by jonathanalevi** [comment]
> > it's still saying quicksilver "will not be compatible with future versions of macOS"
> 
> @jonathanalevi Where do you see this? Is it just on installation?

First time I boot the app up after migration to M1 Mac.

**Comment by skurfer** [comment]
> First time I boot the app up after migration to M1 Mac.

Oh, I suppose any catalog entry that parses HTML Links from a file will also call the script.

**Comment by esoel** [comment]
> We're working on it.
> 
> If anybody wants to give a shot to my first attempt: https://github.com/n8henrie/Quicksilver/releases/download/1.6.1/Quicksilver.1.6.1.dmg -- note that this is not yet merged into quicksilver/Quicksilver, coming from my fork while I work out bugs in the GitHub Actions setup.

Could you please also share the fork/branch you are working on? I might be able to help, I should have some free time over Christmas (no promises...). Thanks

**Comment by n8henrie** [comment]
> Could you please also share the fork/branch you are working on? I might be able to help, I should have some free time over Christmas (no promises...). Thanks

I think the aforementioned changes were merged [here](https://github.com/quicksilver/Quicksilver/pull/2573).

A few others that are not M1 specific but in-progress:

- <https://github.com/quicksilver/Quicksilver/pull/2577>
- <https://github.com/quicksilver/Quicksilver/pull/2582>

**Comment by skurfer** [comment]
I think the majority of the work now will be recompiling all the plug-ins and bumping their versions. Not difficult in most cases, just tedious.

@esoel If you know anything about NIBs, you could start cleaning them up so the preferences and results viewer don’t show random black or blue backgrounds. 🤷🏻‍♂️

**Comment by juanfal** [comment]
> We're working on it.
> 
> If anybody wants to give a shot to my first attempt: https://github.com/n8henrie/Quicksilver/releases/download/1.6.1/Quicksilver.1.6.1.dmg -- note that this is not yet merged into quicksilver/Quicksilver, coming from my fork while I work out bugs in the GitHub Actions setup.

In OSX 12.2 beta (21D5025f), it rejects it telling it is damaged

<img width="420" alt="Captura de pantalla 2021-12-25 a las 12 05 56" src="https://user-images.githubusercontent.com/895596/147383637-4cab7f8c-e01c-418a-89f8-0c3cc0e82b7b.png">

**Comment by skurfer** [comment]
Last night, I built a Universal copy of QS and started building Universal versions of some plug-ins. I started with the ones I care about or thought would be easy to deal with. (See the list below.)

No version bumps or anything yet. Just trying to get them to build. The next steps are:

* Keep running the Intel build with the new plug-ins to make sure they work as expected
* Start bumping versions and releasing Universal versions of the plug-ins
* Switch to a Universal build of QS and make sure things still work as expected
* Deal with the more difficult plug-ins
* Try to fix any issues that come up (like weird colors in the UI)
* Release a Universal build of QS

So far, I’ve rebuilt…

* Audio Plugin
* Calculator Plugin
* Clipboard Plugin
* Command Line Tool
* Dash Plugin
* Event Triggers
* Extra Scripts Plugin
* FaceTime Plugin
* File Attribute Plugin
* File Compression Plugin
* Keyboard Triggers
* MailMate Plugin
* Mouse Triggers
* Networking Plugin
* Nostromo
* Notification Hub
* OnePassword
* Path Finder Plugin
* Process Manipulation Plugin
* QSAddressBookPlugin
* Remote Hosts Plugin
* Safari Plugin
* Services Menu Plugin
* Shelf Plugin
* Spotlight Plugin
* Text Manipulation
* UI Access
* Viscosity Plugin
* Web Search Plugin
* zoom.us Plugin

**Comment by skurfer** [comment]
PayPal is probably the best option for now. https://qsapp.com/donate.php

Donations don’t really impact development because they aren’t large enough for someone to quit their job, but they do help pay for hosting, Apple Dev memberships, etc.
